java/src/org/openqa/selenium/bidi/BiDiProvider.java

```diff
@@ -44,18 +44,32 @@ public Class<HasBiDi> getDescribedInterface() {
 
   @Override
   public HasBiDi getImplementation(Capabilities caps, ExecuteMethod executeMethod) {
+    return new HasBiDi() {
+      private volatile Optional<BiDi> biDi;
 
-    URI wsUri = getBiDiUrl(caps).orElseThrow(() -> new BiDiException("BiDi not supported"));
+      @Override
+      public Optional<BiDi> maybeGetBiDi() {
+        if (biDi == null) {
+          synchronized (this) {
+            if (biDi == null) {
+              URI wsUri =
+                  getBiDiUrl(caps).orElseThrow(() -> new BiDiException("BiDi not supported"));
 
-    HttpClient.Factory clientFactory = HttpClient.Factory.createDefault();
-    ClientConfig wsConfig = ClientConfig.defaultConfig().baseUri(wsUri);
-    HttpClient wsClient = clientFactory.createClient(wsConfig);
-    Connection connection = new Connection(wsClient, wsUri.toString());
+              HttpClient.Factory clientFactory = HttpClient.Factory.createDefault();
+              ClientConfig wsConfig = ClientConfig.defaultConfig().baseUri(wsUri);
+              HttpClient wsClient = clientFactory.createClient(wsConfig);
+              Connection connection = new Connection(wsClient, wsUri.toString());
 
-    return () -> Optional.of(new BiDi(connection));
+              biDi = Optional.of(new BiDi(connection));
+            }
+          }
+        }
+        return biDi;
+      }
+    };
   }
 
-  private Optional<URI> getBiDiUrl(Capabilities caps) {
+  private static Optional<URI> getBiDiUrl(Capabilities caps) {
```

Comment:
Why is this static? Am I missing something?

Reply:
It doesn't really matter. It's a private method, and it doesn't depend on any fields. 
It can be either static or non-static.

Reply:
Then let's remove the `static` keyword.